File: src/VCFGeno.cpp
#include "../include/VCFGeno.h"
#include "../include/common.h"

using namespace std;


//////////////////// VCFGenoBase ////////////////////

VCFGeno *VCFGenoBase::copy() const {
	vector<string>	new_samples(samples.size());
	std::copy(samples.begin(), samples.end(), new_samples.begin());
	vector<GenoRecord *>	new_records(size());
	for(size_t i = 0; i < size(); ++i) {
		new_records[i] = get_record(i)->copy();
	}
	return new VCFGeno(new_samples, new_records, vcf);
}

vector<GenoRecord *> VCFGenoBase::get_geno_records() const {
	vector<GenoRecord *>	records;
	for(size_t i = 0; i < size(); ++i)
		records.push_back(get_record(i));
	return records;
}

vector<size_t> VCFGenoBase::extract_columns(const STRVEC& samples) const {
	map<string, size_t>	dic;	// { sample: column }
	for(size_t i = 0; i < this->samples.size(); ++i)
		dic[this->samples[i]] = i;
	
	vector<size_t>	columns;
	for(auto p = samples.begin(); p != samples.end(); ++p) {
		auto	q = dic.find(*p);
		if(q != dic.end())
			columns.push_back(q->second);
		else
			columns.push_back(string::npos);
	}
	return columns;
}

vector<int> VCFGenoBase::extract_sample_genotypes(size_t sample_index) const {
	vector<int>	gts;
	for(size_t j = 0; j < size(); ++j) {
		const GenoRecord	*record = get_record(j);
		const int	gt = record->get_geno(sample_index);
		gts.push_back(gt);
	}
	return gts;
}

vector<int> VCFGenoBase::clip_raw_haplotype(size_t sample_index,
														int side) const {
	vector<int>	hap;
	for(size_t j = 0; j < size(); ++j) {
		const GenoRecord	*record = get_record(j);
		const int	gt = record->get_geno(sample_index);
		hap.push_back((gt >> side) & 1);
	}
	return hap;
}

vector<size_t> VCFGenoBase::map_samples_to_reference_columns() const {
	// Build a dictionary: reference sample name -> column index in VCF
	// (+9 because VCF sample columns start after the first 9 fixed fields)
	map<string, size_t>	dic_col;
	const STRVEC&	ref_samples = vcf->get_samples();
	for(size_t i = 0; i != ref_samples.size(); ++i) {
		dic_col[ref_samples[i]] = i + 9;
	}
	
	// Prepare result vector initialized with 0
	// (0 means the sample does not exist in the reference VCF)
	vector<size_t>	columns(samples.size(), 0);
	// For each sample in this VCF, look up its column index in the reference
	for(size_t i = 0; i < samples.size(); ++i) {
		auto	q = dic_col.find(samples[i]);
		if(q != dic_col.end())
			columns[i] = q->second;
	}
	return columns;
}

void VCFGenoBase::write(ostream& os, bool with_header) const {
	if(with_header) {
		const auto	header = vcf->trim_header(samples);
		for(auto p = header.begin(); p != header.end(); ++p)
			Common::write_tsv(*p, os);
	}
	
	const vector<size_t>	columns = map_samples_to_reference_columns();
	for(size_t i = 0; i < this->size(); ++i) {
		const GenoRecord	*record = this->get_record(i);
		record->write(vcf->get_record(i), columns, os);
	}
}

VCFGeno *VCFGenoBase::extract_by_samples(const STRVEC& samples) const {
	const auto	cs = extract_columns(samples);
	vector<GenoRecord *>	new_records;
	for(size_t i = 0; i < this->size(); ++i) {
		GenoRecord	*record = this->get_record(i);
		const ll	pos = record->get_pos();
		vector<int>	geno;
		for(auto p = cs.begin(); p != cs.end(); ++p) {
			geno.push_back(record->get_geno(*p));
		}
		GenoRecord	*new_record = new GenoRecord(pos, geno);
		new_records.push_back(new_record);
	}
	return new VCFGeno(samples, new_records, this->vcf);
}


//////////////////// VCFGeno ////////////////////

VCFGeno::~VCFGeno() {
	Common::delete_all(records);
}

vector<vector<int>> VCFGeno::create_ref_haps() const {
	const size_t	 N = num_samples() * 2;
	vector<vector<int>>	ref_haps(N, vector<int>(this->size(), 0));
	for(size_t i = 0; i < N; ++i) {
		for(size_t j = 0; j < this->size(); ++j)
			ref_haps[i][j] = (records[j]->get_geno(i>>1) >> (i&1)) & 1;
	}
	return ref_haps;
}

VCFGeno *VCFGeno::convert(const VCFSmall *vcf) {
	vector<GenoRecord *>	new_records;
	for(size_t i = 0; i < vcf->size(); ++i) {
		const VCFRecord	*record = vcf->get_record(i);
		const auto&	v = record->get_v();
		const ll	pos = record->pos();
		vector<int>	new_geno;
		for(auto p = v.begin()+9; p != v.end(); ++p)
			new_geno.push_back(Genotype::all_gt_to_int(*p));
		auto	*new_record = new GenoRecord(pos, new_geno);
		new_records.push_back(new_record);
	}
	return new VCFGeno(vcf->get_samples(), new_records, vcf);
}

VCFGeno *VCFGeno::join_core(const vector<const VCFGenoBase *>& vcfs,
							const SampleIndexMap& dic, const STRVEC& samples) {
	vector<tuple<string, const VCFGenoBase *, size_t>>	cols;
	for(auto p = samples.begin(); p != samples.end(); ++p) {
		auto	q = dic.find(*p);
		if(q != dic.end())
			cols.push_back(make_tuple(*p, q->second.first, q->second.second));
	}
	
	STRVEC	new_samples;
	for(auto p = cols.begin(); p != cols.end(); ++p)
		new_samples.push_back(get<0>(*p));
	
	vector<GenoRecord *>	new_records;
	for(size_t i = 0; i < vcfs.front()->size(); ++i) {
		const ll	pos = vcfs[0]->get_record(i)->get_pos();
		vector<int>	geno;
		for(auto p = cols.begin(); p != cols.end(); ++p) {
			const auto	*vcf = get<1>(*p);
			const size_t	k = get<2>(*p);
			const auto	*record = vcf->get_record(i);
			const int	gt = record->get_geno(k);
			geno.push_back(gt);
		}
		new_records.push_back(new GenoRecord(pos, geno));
	}
	const VCFSmall	*ref_vcf = vcfs[0]->get_ref_vcf();
	return new VCFGeno(new_samples, new_records, ref_vcf);
}

VCFGeno *VCFGeno::join(const vector<const VCFGenoBase *>& vcfs,
											const STRVEC& samples) {
	// Register all samples with their corresponding VCF and sample index.
	SampleIndexMap	dic;
	for(auto p = vcfs.begin(); p != vcfs.end(); ++p) {
		const VCFGenoBase	*vcf = *p;
		const STRVEC&	ss = vcf->get_samples();
		for(size_t i = 0; i < ss.size(); ++i)
			dic[ss[i]] = make_pair(vcf, i);
	}
	
	return VCFGeno::join_core(vcfs, dic, samples);
}

VCFGeno *VCFGeno::join(const VCFGenoBase *vcf1, const VCFGenoBase *vcf2,
														const STRVEC& samples) {
	if(vcf1 == NULL)
		return vcf2->copy();
	
	vector<const VCFGenoBase *>	vcfs { vcf1, vcf2 };
	return VCFGeno::join(vcfs, samples);
}

VCFGeno *VCFGeno::create_by_two_vcfs(const VCFGenoBase *vcf1,
											const VCFSmall *vcf2,
											const STRVEC& samples) {
	// assume that samples are [mat, pat, prog1, prog2, ...]
	const vector<size_t>	columns1 = vcf1->extract_columns(samples);
	const vector<size_t>	columns2 = vcf2->extract_columns(samples);
	vector<GenoRecord *>	new_records;
	for(size_t i = 0; i < vcf1->size(); ++i) {
		const auto	*record1 = vcf1->get_record(i);
		const auto	*record2 = vcf2->get_record(i);
		vector<int>	geno;
		for(size_t j = 0; j < samples.size(); ++j) {
			if(columns1[j] != string::npos)
				geno.push_back(record1->get_geno(columns1[j]));
			else if(columns2[j] != string::npos)
				geno.push_back(
					Genotype::all_gt_to_int(record2->get_gt(columns2[j]-9)));
			else
				geno.push_back(Genotype::NA);
		}
		auto	*new_record = new GenoRecord(record1->get_pos(), geno);
		new_records.push_back(new_record);
	}
	return new VCFGeno(samples, new_records, vcf2);
}

VCFGeno *VCFGeno::extract_samples(const STRVEC& samples, const VCFSmall *vcf) {
	const auto	cs = vcf->extract_columns(samples);
	vector<GenoRecord *>	new_records;
	for(size_t i = 0; i < vcf->size(); ++i) {
		const VCFRecord	*record = vcf->get_record(i);
		const ll	pos = record->pos();
		vector<int>	new_geno;
		for(auto p = cs.begin(); p != cs.end(); ++p) {
			const auto	c = *p;
			const int	gt = c == string::npos ? Genotype::NA :
									Genotype::all_gt_to_int(record->get_v()[c]);
			new_geno.push_back(gt);
		}
		auto	*new_record = new GenoRecord(pos, new_geno);
		new_records.push_back(new_record);
	}
	return new VCFGeno(samples, new_records, vcf);
}

File: python/VCFGeno.py
from __future__ import annotations

# coding: utf-8
# VCFGeno.py
# 基本的にGenotypeのみ持つ

from itertools import count
from abc import ABC, abstractmethod
from typing import TextIO

from GenoRecord import GenoRecord
from VCF import VCFSmall, VCFRecord
from Genotype import Genotype


#################### VCFGenoBase ####################

class VCFGenoBase(ABC):
	def __init__(self, samples: list[str], vcf: VCFSmall) -> None:
		self.samples: list[str] = samples
		self.vcf: VCFSmall = vcf
	
	@abstractmethod
	def __len__(self) -> int:
		pass
	
	@abstractmethod
	def get_record(self, i: int) -> GenoRecord:
		pass
	
	def get_ref_vcf(self) -> VCFSmall:
		return self.vcf
	
	def get_samples(self) -> list[str]:
		return self.samples
	
	def num_samples(self) -> int:
		return len(self.samples)
	
	def extract_columns(self, samples: list[str]) -> list[int]:
		dic = dict(zip(self.samples, count()))
		return [ dic.get(sample, -1) for sample in samples ]
	
	def clip_raw_haplotype(self, sample_index: int, side: int) -> list[int]:
		hap: list[int] = []
		for j in range(len(self)):
			record = self.get_record(j)
			h = record.get_allele(sample_index, side)
			hap.append(h)
		return hap
	
	def write(self, out: TextIO, with_header: bool=True) -> None:
		if with_header:
			self.vcf.write_header(out)
		for i in range(len(self)):
			self.get_record(i).write(self.vcf.records[i], out)
	
	def extract_by_samples(self, samples: list[str]) -> VCFGeno:
		cs = self.extract_columns(samples)
		new_records: list[GenoRecord] = []
		for i in range(len(self)):
			record = self.get_record(i)
			pos = record.pos
			geno = [ record.geno[c] for c in cs ]
			new_record = GenoRecord(pos, geno)
			new_records.append(new_record)
		return VCFGeno(samples, new_records, self.vcf)


#################### VCFGeno ####################

class VCFGeno(VCFGenoBase):
	def __init__(self, samples: list[str],
						records: list[GenoRecord], vcf: VCFSmall) -> None:
		super().__init__(samples, vcf)
		self.records: list[GenoRecord] = records
	
	def __len__(self) -> int:
		return len(self.records)
	
	def get_record(self, i: int) -> GenoRecord:
		return self.records[i]
	
	@staticmethod
	def extract_samples(samples: list[str], vcf: VCFSmall) -> VCFGeno:
		cs = vcf.extract_columns(samples)
		new_records: list[GenoRecord] = []
		for i in range(len(vcf)):
			record = vcf.get_record(i)
			pos = int(record.v[1])
			new_v = [ Genotype.all_gt_to_int(record.v[c]) for c in cs ]
			new_record = GenoRecord(pos, new_v)
			new_records.append(new_record)
		return VCFGeno(samples, new_records, vcf)
	
	# samplesの順番で結合
	@staticmethod
	def join(vcfs: list[VCFGenoBase], samples: list[str]) -> VCFGeno:
		dic = { }
		for vcf in vcfs:
			# sに'0'があっても問題ない
			for c, s in enumerate(vcf.samples):
				dic[s] = (vcf, c)
		cols = []
		for s in samples:
			if s in dic:
				cols.append((s,) + dic[s])
		
		new_samples = [ s for s, vcf, c in cols ]
		new_records = []
		for i in range(len(vcfs[0])):
			geno = []
			for s, vcf, c in cols:
				geno.append(vcf.get_record(i).geno[c])
			new_records.append(GenoRecord(vcfs[0].get_record(i).pos, geno))
		return VCFGeno(new_samples, new_records, vcfs[0].vcf)
	
	# vcf1にあるsampleはvcf1から、
	# そうでないsampleはvcf2からGenotypeを取って新たなVCFを作る
	@staticmethod
	def create_by_two_vcfs(vcf1: VCFGenoBase, vcf2: VCFSmall,
										samples: list[str]) -> VCFGeno:
		# samplesは[mat, pat, prog1, ...]の前提
		columns1 = vcf1.extract_columns(samples)
		columns2 = vcf2.extract_columns(samples)
		new_records: list[GenoRecord] = []
		for i in range(len(vcf1)):
			record1: GenoRecord = vcf1.get_record(i)
			record2: VCFRecord = vcf2.get_record(i)
			geno: list[int] = []
			for j in range(len(samples)):
				if columns1[j] != -1:
					geno.append(record1.geno[columns1[j]])
				elif columns2[j] != -1:
					geno.append(Genotype.all_gt_to_int(record2.v[columns2[j]]))
				else:	# どちらのVCFにもそのサンプルは無い
					geno.append(Genotype.NA)
			new_record = GenoRecord(record1.pos, geno)
			new_records.append(new_record)
		return VCFGeno(samples, new_records, vcf2)
